README.md
![Dart CI](https://github.com/jbytecode/boxplot/workflows/Dart%20CI/badge.svg)

A Boxplot library for Dart developers.

Created by Matletik - Mathematical Manipulatives and Interactive Learning Tools

http://www.matletik.com

[Developer Web Site](http://www.matletik.com)

[Source Code](https://github.com/jbytecode/boxplot)

![Screenshot](https://github.com/jbytecode/boxplot/raw/master/images/screenshot.png)

## Usage

A simple usage example:

```dart
import 'dart:html';
import 'package:boxplot/box_plot.dart';

void main() {
  DivElement mydiv = querySelector("#mydiv");
  BoxPlot boxplot = new BoxPlot(500, 200);
  boxplot.setQuantiles(0, 30, 60, 86, 100);
  mydiv.append(boxplot.getDrawableComponent());


  RangeInputElement myrange = querySelector("#myrange");
  myrange.onInput.listen((event) {
    boxplot.setValue(double.parse(myrange.value));
    boxplot.drawComponent();
  });
}
```

```html
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>Matletik Boxplot for Dart - Example</title>
    <script src = "boxplot_example.dart.js"></script>

    <style>
        div{
            background-color: rgba(255, 255, 255, 0.35);
            font-size: 16px;
            margin: 0 auto;
            valign: middle;
            text-align: center;
        }
        input[type=range]{
            width: 100%;
        }
    </style>
</head>
<body>
<fieldset>
    <caption>Value on the boxplot</caption>
    <input type="range" min="0" max="100" value="65" id="myrange"/>
</fieldset>
<br/>
<div id="mydiv"></div>
</body>
</html>
```

## Features and bugs

Please file feature requests and bugs at the [issue tracker][tracker].

[tracker]: http://example.com/issues/replaceme


CHANGELOG.md
## 1.0.5

- Quantiles can be set manually as well as automatically. 
A basic algorithm is used to estimate 25th and 75th quantiles.
- Axes limits and labels are now changing automatically.

## 1.0.3

- Package description in more detail.

## 1.0.2

- Code refactoring and Dart docs.

## 1.0.1

- Code refactoring and Dart docs.

## 1.0.0

- Initial version, created by Matletik


test/boxplot_test.dart
import 'package:boxplot/src/stats.dart';
import 'package:test/test.dart';

void main() {
  group('A group of tests', () {
    setUp(() {});

    test('Minimum', () {
      List<double> arr = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, -1];
      expect(-1, Stats.min(arr));
    });

    test('Maximum', () {
      List<double> arr = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, -1];
      expect(10, Stats.max(arr));
    });

    test('Median - Number of elements is even', () {
      List<double> arr = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
      expect(Stats.median(arr), 5.5);
    });

    test('Median - Number of elements is odd', () {
      List<double> arr = [1, 2, 3, 4, 5, 6, 7, 8, 9];
      expect(Stats.median(arr), 5.0);
    });

    test('q1', () {
      List<double> arr = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11];
      expect(Stats.q1(arr), 3);
    });

    test('q3', () {
      List<double> arr = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11];
      expect(Stats.q3(arr), 9);
    });
  });
}


lib/boxplot.dart
/*
 *
 boxplot, A Dart library for Box plot tool.
 Copyright (C) 2020-2021  Matletik - Mathematical Manipulatives and Interactive Learning Tools
 This program is free software: you can redistribute it and/or modify
 it under the terms of the GNU Lesser General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 any later version.
 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
 You should have received a copy of the GNU Lesser General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 *
 * Matletik - Mathematical Manipulatives and Interactive Learning Tools
 * http://www.matletik.com
 */
library boxplot;

export 'src/box_plot.dart';
export 'src/gui_component.dart';


lib/src/stats.dart
class Stats {
  /// fit a value x in [min, max] to [0,100]
  static arrange(double min, double max, double value) {
    return 100 * ((value - min) / (max - min));
  }

  /// fit a value x in [0, 100] to [min, max]
  static rearrange(double min, double max, double value) {
    return ((max - min) * (value) / 100.0) + min;
  }

  /// minimum of the list
  static double min(List<double> arr) {
    double min = arr[0];
    arr.forEach((double element) {
      if (element < min) {
        min = element;
      }
    });
    return min;
  }

  /// maximum of the list
  static double max(List<double> arr) {
    double max = arr[0];
    arr.forEach((double element) {
      if (element > max) {
        max = element;
      }
    });
    return max;
  }

  /// A basic estimator for median
  static double median(List<double> arr) {
    List<double> temp = List.from(arr);
    temp.sort((double a, double b) {
      return a.compareTo(b);
    });
    int length = temp.length;
    if (length % 2 == 1) {
      return temp[(length / 2).ceil() - 1];
    } else {
      return (temp[(length / 2).toInt() - 1] +
              temp[((length / 2) + 1).toInt() - 1]) *
          0.5;
    }
  }

  /// A basic estimator for 25th percentile
  static double q1(List<double> arr) {
    double mymedian = Stats.median(arr);
    List<double> temp = arr.where((double element) {
      return element < mymedian;
    }).toList();
    return Stats.median(temp);
  }

  /// A basic estimator for 75th percentile
  static double q3(List<double> arr) {
    double mymedian = Stats.median(arr);
    List<double> temp = arr.where((double element) {
      return element > mymedian;
    }).toList();
    return Stats.median(temp);
  }
}


example/boxplot_example.dart
import 'dart:html';
import 'package:boxplot/boxplot.dart';

void main() {
  DivElement mydiv = querySelector("#mydiv");
  /*
  Creating a boxplot with the dimension 500 x 200 in pixels
   */
  BoxPlot boxplot = new BoxPlot(500, 200);

  /*
  Quantiles can be set manually
   */
  boxplot.setQuantiles(
      /* min */
      50,
      /* 25th percentile */ 60,
      /* median */ 70,
      /* 75th percentile */ 80,
      /* max */ 200);

  /*
  Appending drawable component into a div element
   */
  mydiv.append(boxplot.getDrawableComponent());

  /*
  or, quantiles can be estimated using a dataset:
   */
  //boxplot.setQuantilesUsingData(aListOfDoubleValues);

  RangeInputElement myrange = querySelector("#myrange");
  myrange.onInput.listen((event) {
    boxplot.setValue(double.parse(myrange.value));
    boxplot.drawComponent();
  });
}


lib/src/box_plot.dart
/*
 *
 boxplot, A Dart library for Box plot tool.
 Copyright (C) 2020-2021  Matletik - Mathematical Manipulatives and Interactive Learning Tools
 This program is free software: you can redistribute it and/or modify
 it under the terms of the GNU Lesser General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 any later version.
 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
 You should have received a copy of the GNU Lesser General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 *
 * Matletik - Mathematical Manipulatives and Interactive Learning Tools
 * http://www.matletik.com
 */

import 'dart:html';
import 'dart:math' as Math;
import 'package:boxplot/src/stats.dart';

import 'gui_component.dart';

/// Canvas based BoxPlot GUI Component
class BoxPlot extends GuiComponent {
  bool showPoint = true;

  /// Width and height of component in pixels
  int width, height;

  /// Canvas element to draw
  CanvasElement canvas;

  /// The 2D rendering Context
  CanvasRenderingContext2D g2;

  /// Basic statistics to draw boxplot
  double min, max, q1, q2, q3, value;

  /// Virtual cartesian dimensions for drawing
  double minx, miny, maxx, maxy;

  /// Color of the boxplot
  String boxPlotcolor = '#fed8b1';

  /// Color of the ball which defines a single observation given by [value]
  String ballColor = 'rgba(0, 0, 255, 0.50)';

  /// Font size and name for axes labels
  String font = '14px Monospace';

  /// True if mouse entered on the plot
  bool mouseEntered = false;

  /// Mouse event for controlling mouse entering in the plot
  MouseEvent mouseEvent;

  /// The constructor
  BoxPlot(int widthInPixels, int heightInPixels) {
    width = widthInPixels;
    height = heightInPixels;
    canvas = CanvasElement();
    canvas.style.border = '1px solid black';
    canvas.width = widthInPixels;
    canvas.height = heightInPixels;
    g2 = canvas.getContext('2d');
    // Dimension of the virtual cartesian
    setDimensions(-10, 110, -11, 11);
    canvas.onMouseEnter.listen((Event e) {
      mouseEntered = true;
    });
    canvas.onMouseOut.listen((Event e) {
      mouseEntered = false;
      drawComponent();
    });
    canvas.onMouseMove.listen((MouseEvent e) {
      mouseEvent = e;
      drawComponent();
    });
    setDefaultQuantiles();
  }

  /// Sets the [value] point is visible or hidden
  void setShowPoint(bool show) {
    showPoint = show;
  }

  /// Gets the visibility state of [value] point
  bool getShowPoint() {
    return showPoint;
  }

  /// Sets the default statistics
  void setDefaultQuantiles() {
    min = 0;
    max = 100;
    q1 = 25;
    q2 = 50;
    q3 = 75;
    value = 65;
  }

  /// Returns the statistics as a Map
  Map<String, double> getDefaultQuantiles() {
    Map<String, double> quantiles = Map<String, double>();
    quantiles['min'] = min;
    quantiles['max'] = max;
    quantiles['q1'] = q1;
    quantiles['q2'] = q2;
    quantiles['q3'] = q3;
    quantiles['value'] = value;
    return quantiles;
  }

  /// Sets the color of box plot
  void setBoxPlotColor(String color) {
    boxPlotcolor = color;
  }

  /// Returns the color of box plot
  String getBoxPlotColor() {
    return boxPlotcolor;
  }

  /// Changes the ball of [value] color
  void setBallColor(String colorCode) {
    ballColor = colorCode;
  }

  /// Gets the color of ball which in the coordinate of [value]
  String getBallColor() {
    return ballColor;
  }

  /// Changes the axis font
  void setFont(String font) {
    this.font = font;
  }

  /// Returns the current axis font
  String getFont() {
    return font;
  }

  /// Sets the current location of the ball displayed in the plot
  void setValue(double value) {
    this.value = value;
  }

  /// Returns the coordinate of ball
  double getValue() {
    return value;
  }

  /// Sets the basic statistics as a list
  /// Elements are ordered as [min], [q1], [q2], [q3], and [max]
  void setQuantilesWithList(List<double> values) {
    min = values[0];
    q1 = values[1];
    q2 = values[2];
    q3 = values[3];
    max = values[4];
  }

  /// Sets the basic statistics as single arguments for each
  void setQuantiles(double min, double q1, double q2, double q3, double max) {
    setQuantilesWithList([min, q1, q2, q3, max]);
  }

  /// Sets the basic statistics using data
  void setQuantilesUsingData(List<double> arr) {
    setQuantilesWithList([
      Stats.min(arr),
      Stats.q1(arr),
      Stats.median(arr),
      Stats.q3(arr),
      Stats.max(arr)
    ]);
  }

  /// Sets dimensions of the virtual cartesian
  void setDimensions(double minx, double maxx, double miny, double maxy) {
    this.minx = minx;
    this.maxx = maxx;
    this.miny = miny;
    this.maxy = maxy;
  }

  /// Translate the y coordinate from virtual cartesian to real coordinates.
  double translateY(double y) {
    return -(miny + y) * height / (maxy - miny);
  }

  /// Translate the x coordinate from virtual cartesian to real coordinates.
  double translateX(double x) {
    return (x - minx) * width / (maxx - minx);
  }

  /// Translate the x coordinate from real coordinates to virtual cartesian.
  double retranslateX(double x) {
    return ((x * (maxx - minx)) + (width * minx)) / width;
  }

  /// Translate the y coordinate from real coordinates to virtual cartesian.
  double retranslateY(double y) {
    return (-1) * (y * (maxy - miny) + (miny * height)) / (height);
  }

  /// Refreshes the content of the component
  void drawComponent() {
    double max = 100.0;
    double min = 0.0;
    double q1 = Stats.arrange(this.min, this.max, this.q1);
    double q2 = Stats.arrange(this.min, this.max, this.q2);
    double q3 = Stats.arrange(this.min, this.max, this.q3);
    double value = Stats.arrange(this.min, this.max, this.value);

    g2.lineWidth = 1;
    g2.fillStyle = boxPlotcolor;
    g2.strokeStyle = 'black';

    g2.resetTransform();
    g2.clearRect(0, 0, width, height);

    // Draw Axes
    g2.beginPath();
    g2.moveTo(translateX(0), translateY(2));
    g2.lineTo(translateX(100), translateY(2));
    g2.closePath();
    g2.stroke();

    //Draw minx
    g2.beginPath();
    g2.moveTo(translateX(min), translateY(-2));
    g2.lineTo(translateX(min), translateY(6));
    g2.closePath();
    g2.stroke();

    //Draw maxx
    g2.beginPath();
    g2.moveTo(translateX(max), translateY(-2));
    g2.lineTo(translateX(max), translateY(6));
    g2.closePath();
    g2.stroke();

    // Box
    g2.beginPath();
    g2.moveTo(translateX(q1), translateY(-2));
    g2.lineTo(translateX(q1), translateY(6));
    g2.lineTo(translateX(q3), translateY(6));
    g2.lineTo(translateX(q3), translateY(-2));
    g2.closePath();
    g2.fillStyle = 'white';
    g2.fill();
    g2.fillStyle = boxPlotcolor;
    g2.fill();
    g2.strokeStyle = "black";
    g2.stroke();

    // Median
    g2.beginPath();
    g2.moveTo(translateX(q2), translateY(-2));
    g2.lineTo(translateX(q2), translateY(6));
    g2.closePath();
    g2.stroke();

    g2.font = '14px Monospace';
    g2.fillStyle = 'black';

    // Axis labels
    for (double i = min; i <= max; i += 10.0) {
      g2.fillText(Stats.rearrange(this.min, this.max, i).toInt().toString(),
          translateX(i - 2), translateY(-10));
    }

    //ticks
    double linewidth = 1.0;
    for (int i = 0; i <= 100; i += 10) {
      g2.beginPath();
      if ((i % 10) == 0) {
        linewidth = 1.0;
      } else {
        linewidth = 0.5;
      }
      g2.moveTo(translateX(i.toDouble()), translateY(-5.5 - linewidth));
      g2.lineTo(translateX(i.toDouble()), translateY(-5.5 + linewidth));
      g2.closePath();
      g2.stroke();
    }

    // the line above the numbers
    g2.beginPath();
    g2.moveTo(translateX(0), translateY(-4.6));
    g2.lineTo(translateX(100), translateY(-4.6));
    g2.closePath();
    g2.strokeStyle = 'black';
    g2.lineWidth = 0.5;
    g2.stroke();

    //Where is mouse
    if (mouseEntered) {
      double x = mouseEvent.client.x - canvas.getBoundingClientRect().left;
      double y = mouseEvent.client.y - canvas.getBoundingClientRect().top;
      g2.beginPath();
      g2.rect(x, y, 70, 15);
      g2.closePath();
      g2.fillStyle = 'rgba(0,0,255,0.25)';
      g2.strokeStyle = 'black';
      g2.fill();
      g2.stroke();

      g2.font = font;
      g2.fillStyle = 'black';
      g2.fillText(
          (Stats.rearrange(this.min, this.max, retranslateX(x)))
              .toStringAsFixed(2),
          x + 10,
          y + 12);
    }

    // Ball point
    if (showPoint) {
      g2.beginPath();
      window.console.log(value);
      g2.arc(translateX(value), translateY(2), 5, 0.0, 2.0 * Math.pi, false);
      g2.closePath();
      g2.fillStyle = 'white';
      g2.fill();
      g2.fillStyle = ballColor;
      g2.fill();
      g2.strokeStyle = 'black';
      g2.stroke();
    }
  }

  @override
  HtmlElement getDrawableComponent() {
    drawComponent();
    return canvas;
  }
}
